File: ruby/luhn/luhn.rb
class Luhn
  def self.valid?(str)
    new(str).valid?
  end

  def initialize(str)
    @string = str.delete(' ')
  end

  def valid?
    valid_format? && (checksum % 10).zero?
  end

  private

  attr_reader :string

  def valid_format?
    string.match(/^\d{2,}$/)
  end

  def checksum
    string.reverse.chars.each_slice(2).sum do |even, odd|
      even.to_i + (odd.to_i * 2).digits.sum
    end
  end
end

File: javascript/luhn/luhn.js
export class Luhn {
  constructor(inputString) {
    this.inputString = inputString.replace(/\s/g, '');
    this.valid = this.checkValidity();
  }

  checkValidity() {
    if (this.inputString.length === 1) { return false; }

    let sum = 0;
    [...this.inputString].reverse().forEach((element, index) => {
      if (index % 2 !== 0) {
        let doubled = (parseInt(element) * 2);
        sum += doubled > 9 ? doubled - 9 : doubled;
      } else {
        sum += parseInt(element);
      }
    });

    return sum % 10 === 0;
  }
}
